arch/Mips/MipsMapping.c

```diff
@@ -1,1071 +1,454 @@
-	{ MIPS_GRP_NOTINMICROMIPS, "notinmicromips" },
-	{ MIPS_GRP_NOTNACL, "notnacl" },
+	if (doing_mem(MI)) {
+		Mips_set_detail_op_mem_disp(MI, OpNum, Imm);
+		return;
+	}
 
-	{ MIPS_GRP_NOTMIPS32R6, "notmips32r6" },
-	{ MIPS_GRP_NOTMIPS64R6, "notmips64r6" },
-	{ MIPS_GRP_CNMIPS, "cnmips" },
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_IMM;
+	Mips_get_detail_op(MI, 0)->imm = Imm;
+	Mips_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
+	Mips_inc_op_count(MI);
+}
 
-	{ MIPS_GRP_MIPS32, "mips32" },
-	{ MIPS_GRP_MIPS32R6, "mips32r6" },
-	{ MIPS_GRP_MIPS64R6, "mips64r6" },
+static void Mips_set_detail_op_reg(MCInst *MI, unsigned OpNum, mips_reg Reg)
+{
+	if (!detail_is_set(MI))
+		return;
 
-	{ MIPS_GRP_MIPS2, "mips2" },
-	{ MIPS_GRP_MIPS3, "mips3" },
-	{ MIPS_GRP_MIPS3_32, "mips3_32"},
-	{ MIPS_GRP_MIPS3_32R2, "mips3_32r2" },
+	if (doing_mem(MI)) {
+		Mips_set_detail_op_mem_reg(MI, OpNum, Reg);
+		return;
+	}
 
-	{ MIPS_GRP_MIPS4_32, "mips4_32" },
-	{ MIPS_GRP_MIPS4_32R2, "mips4_32r2" },
-	{ MIPS_GRP_MIPS5_32R2, "mips5_32r2" },
+	assert((map_get_op_type(MI, OpNum) & ~CS_OP_MEM) == CS_OP_REG);
 
-	{ MIPS_GRP_GP32BIT, "gp32bit" },
-	{ MIPS_GRP_GP64BIT, "gp64bit" },
-};
-#endif
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_REG;
+	Mips_get_detail_op(MI, 0)->reg = Reg;
+	Mips_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
+	Mips_inc_op_count(MI);
+}
 
-const char *Mips_group_name(csh handle, unsigned int id)
+static void Mips_set_detail_op_operand(MCInst *MI, unsigned OpNum)
 {
-#ifndef CAPSTONE_DIET
-	return id2name(group_name_maps, ARR_SIZE(group_name_maps), id);
-#else
-	return NULL;
-#endif
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		Mips_set_detail_op_imm(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else if (op_type == CS_OP_REG) {
+		Mips_set_detail_op_reg(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else
+		printf("Operand type %d not handled!\n", op_type);
+}
+
+static void Mips_set_detail_op_branch(MCInst *MI, unsigned OpNum)
+{
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		uint64_t Target = (uint64_t)MCInst_getOpVal(MI, OpNum);
+		Mips_set_detail_op_imm(MI, OpNum, Target + MI->address);
+	} else if (op_type == CS_OP_REG) {
+		Mips_set_detail_op_reg(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else
+		printf("Operand type %d not handled!\n", op_type);
+}
+
+static void Mips_set_detail_op_uimm(MCInst *MI, unsigned OpNum)
+{
+	Mips_set_detail_op_imm(MI, OpNum,
+				    MCInst_getOpVal(MI, OpNum));
+}
+
+static void Mips_set_detail_op_uimm_offset(MCInst *MI, unsigned OpNum,
+											unsigned Bits, uint64_t Offset)
+{
+	uint64_t Imm = MCInst_getOpVal(MI, OpNum);
+	Imm -= Offset;
+	Imm &= (1 << Bits) - 1;
+	Imm += Offset;
+	Mips_set_detail_op_imm(MI, OpNum, Imm);
+}
+
+static void Mips_set_detail_op_mem_nanomips(MCInst *MI, unsigned OpNum)
+{
+	if (!detail_is_set(MI) || !doing_mem(MI))
```

Comment:
Better move the `doing_mem()` check to a separated `if` and print a warning if it is false. The function should only be called for memory operands. Not returning silently will ease debugging and discovering bugs a lot.

Reply:
Especially discovering faulty logic after this big refactor.

Reply:
that `doing_mem` should be always false. it's there just in case somebody modifies the code. it could be an assert.

Reply:
`CS_ASSERT` is better in my experience (gets removed in release build). I spent quite some debugging time in the past, because these things were silent.